pyecoregen/ecore.py

```diff
@@ -269,6 +297,44 @@ def create_environment(self, **kwargs):
 
         return environment
 
+    @staticmethod
+    def resolve_all_proxies(root: ecore.EObject):
+        """Resolves all the proxies from an object"""
+        for o in root.eAllContents():
+            if isinstance(o, ecore.EEnum):
+                continue
+            for reference in o.eClass.eAllReferences():
+                values = o.eGet(reference)
+                if not reference.many:
+                    values = [values]
+                for value in values:
+                    if not isinstance(value, ecore.EProxy):
+                        continue
+                    value.eResource  # force proxy resolution
+
+    @staticmethod
+    def load_all_required_resources(root):
+        """
+        Returns a set of all the resources on which depends a model root.
+
+        Args:
+            root: the model root that will be traversed to find dependencies.
+        """
+        rset = root.eResource.resource_set
+        before = {v for _, v in rset.resources.items()}
+        for x in before:
+            EcoreGenerator.resolve_all_proxies(x.contents[0])
+        current = {v for _, v in rset.resources.items()}
+        diff = current - before
+        for x in diff:
+            current |= EcoreGenerator.load_all_required_resources(x.contents[0])
+        return current
+
     def generate(self, model, outfolder):
         with pythonic_names():
             super().generate(model, outfolder)
+            if self.with_dependencies and model.eResource:
+                all_resources = self.load_all_required_resources(model)
+                all_resources.remove(model.eResource)
+                for resource in all_resources:
+                    super().generate(resource.contents[0], outfolder)
```

Comment:
That does not look quite right at first glance. Here is what I understand you are doing:

1. Load all referenced models.
2. Call the base class generation method for each of them.

What I am missing:
* Assuming this class' `generate` function would add other, super-important logic, it would not be applied to the referenced models (as the base class method is called for those only).
* Are transitive references processed or is this not required?

I would have expected that the generate method basically calls its *own* implementation with the referenced models. This way specific behavior is also applied to those models and their references are also processed. But I may be off how references are managed in `pyecore`.

Reply:
The `load_required_resources(...)` methods also loads the transitive metamodels (I will add this in the method documentation), and the same 'generate' is used  for each dependency (including transitive ones) in order to keep the same options through all generations. 

The fact to compute first all the dependencies (with the transitive ones) is to avoid multiple generation of the same artifact. E.g: `A` depends on `B` and `C` and `B` depends on `C` only, in this case, `C` would be generated twice if the dependencies are generated one after the other.

I can create a new `EcoreGenerator` and set it up with the same options than the *caller* generator. Do you think this would be better?

Reply:
I'd still like to see this recursion. Also what would happen if someone else does a similar thing in a derived generator class, calling the above `generate` method multiple times with different models? This would still possibly generate the same submodel multiple times, wouldn't it?

Would it make sense to add an optional argument to `generate` that holds the set of generated models/refs/names, and is initially empty? Like:

```python
def generate(self, model, outfolder, exclude=None):
    if exclude is None:
        exclude = set()
    ....
    super().generate(model, outfolder)
    exclude.add(package)
    ....
    for package in ...:
        self.generate(package, outfolder, exclude)
        exclude.add(package)
```

What do you think?

Reply:
The recusion is in the `load_required_resources(...)` static methods:

```python
for x in diff:
     current |= EcoreGenerator.load_all_required_resources(x.contents[0])
```

Regarding a possible derived generator, indeed, if the `with_dependencies` is set to `True`, and the derived generator asks to generate some transitive dependencies, they will be generated multiple times. As I though it will be a rare case as you have to explicitally activate this option (so if you do, you know what it implies), I let this case on the side.

I guess adding an optional parameter should not be an issue, this will imply the removal of the `load_required_resources(...)` method in order to directly call the `resolve_all_proxies(...)` one.

Reply:
I changed the algorithm so it use a direct recursion as you proposed, indeed, it looks better!